### PR TITLE
Feature/bsn

### PR DESCRIPTION
**Pull request by dunyakirkali**
We needed a way to generate BSN numbers ([burgerservicenummer](http://nl.wikipedia.org/wiki/Burgerservicenummer)) in order to test an application.

So we decided to extend the faker library with the capability to generate valid BSN numbers.

There are tests for it and the README is updated.

### CODE REVIEW

**Comment by dunyakirkali** [comment]
@tolgap feel like helpin' me out with tidying up this PR?

**Comment by Marthyn** [comment]
@dunyakirkali tried something here, what do you think? It's a teeny tiny bit faster :dancers: :

**Comment by dunyakirkali** [comment]
@Marthyn That's more like it. Thanx :)

**Comment by Marthyn** [comment]
np, what was this line for 
`bsn.concat(total % 11)`
Because in your code it didn't do anything to the end result. And it doesn't affect the tests.

**Comment by dunyakirkali** [comment]
To be honest I have no clue, this is actually Tolga's code. I'm just the messenger :)

**Comment by tolgap** [comment]
BSN needs to be divisible by 11. If it's not, the remainder is added to make sure its divisible by 11.

**Comment by Marthyn** [comment]
But you did this, which means nothing happens to bsn.

```
bsn.concat(total % 11)
return bsn.to_i if total % 11 == 0
```

 you'd have to do 

```
bsn  = bsn.concat(total % 11)
return bsn.to_i if total % 11 == 0
```

But also in this case it will generate a new bsn anyways becaus `total % 11` still is not 0

**Comment by tolgap** [comment]
@Marthyn 

```
[3] pry(main)> a = 'hello';
[4] pry(main)> a
=> "hello"
[5] pry(main)> a.concat(' world');
[6] pry(main)> a
=> "hello world"
```

`.concat` works on the same String.

**Comment by Marthyn** [comment]
You're right, my bad :smile: 

Spec should be fixed so without the concatenate it fails.

**Comment by dunyakirkali** [comment]
Who will fix them? :)

**Comment by Marthyn** [comment]
I'll have a look :)

Sent from my iPhone

On Wed, Jun 3, 2015 at 3:30 AM -0700, "Dunya Kirkali" notifications@github.com wrote:
Who will fix them? :)

---

**Comment by Marthyn** [comment]
```
class Bsn < Base
    class << self
      # returns a valid bsn number of length 8 or 9
      def bsn
        loop do
          bsn = ''
          total = (0..8).inject do |total, i|
            random_number = rand(i == 0 ? 2 : 9)
            bsn += random_number.to_s
            total += random_number * (i == 8 ? -1 : (9 - i))
          end
          return bsn.to_i if total % 11 == 0
        end
      end
    end
  end
```

or

```
class Bsn < Base
    class << self
      # returns a valid bsn number of length 8 or 9
      def bsn
        loop do
          bsn = ''
          total = (0..8).inject do |total, i|
            random_number = random_for(i)
            bsn += random_number.to_s
            total += random_number * (i == 8 ? -1 : (9 - i))
          end
          return bsn.to_i if total % 11 == 0
        end
      end

      private

      def random_for(i)
        rand(i == 0 ? 2 : 9)
      end
    end
  end
```

**Comment by dunyakirkali** [comment]
I prefer the second version. The **more functions** the **better**! Find you not?

**Comment by Marthyn** [comment]
I do but i was inspired by a talk about writing the shortest code possible :dancer:

**Comment by dunyakirkali** [comment]
Oh no.
I would go for 'The shortest, most legible code possible' :)

**Comment by Marthyn** [comment]
How bout this :smile_cat:

**Comment by dunyakirkali** [comment]
That looks very good :wine_glass:

**Comment by Marthyn** [comment]
Oh also! In the spec the size of an Integer is called, which is always 8, (8 bits i presume) 

```
irb(main):001:0> 1.size
=> 8
```

so to test the length of the bsn

```
bsn.to_s.length
```

or

```
Math.log10(bsn).to_i + 1
```

**Comment by Marthyn** [comment]
very interesting exercise this ;)

**Comment by dunyakirkali** [comment]
Oh sh*%$. That .size is my fault. Stupid.

P.S. this is definitely good exercise. Reminds me of these codekata exercices.

File: test/test_faker_bsn.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+require File.expand_path(File.dirname(__FILE__) + '/test_helper.rb')
+
+class TestFakerBsn < Test::Unit::TestCase
+  def test_bsn_length
+    bsn = Faker::Bsn.bsn
+    assert [8,9].include? bsn.to_s.length
+  end
+
+  def test_bsn_validity
+    100.times do
```
Comment (Marthyn) [inline test/test_faker_bsn.rb:10]: i added 100 times because its a random number because this method is not idempotent.

Comment (dunyakirkali) [inline test/test_faker_bsn.rb:10]: That probably doesn't slow the test suite all that much, right?

Comment (Marthyn) [inline test/test_faker_bsn.rb:10]:
```
Finished in 0.019484 seconds.

2 tests, 101 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
```

^ with 100.times
VS
v without 100.times 

```
Finished in 0.001817 seconds.

2 tests, 2 assertions, 0 failures, 0 errors, 0 pendings, 0 omissions, 0 notifications
100% passed
```

Comment (dunyakirkali) [inline test/test_faker_bsn.rb:10]: I would call it 'acceptable'. :) u guys agree?

**Comment by dunyakirkali** [comment]
@tolgap can you add your latest fixes to this branch?

**Comment by dunyakirkali** [comment]
Jenkins test this please

**Comment by dunyakirkali** [comment]
Jenkins test this please

**Comment by dunyakirkali** [comment]
Jenkins test this please

**Comment by dunyakirkali** [comment]
Jenkins test this please

**Pull request closed without merging**